rbuf.go

```diff
@@ -226,6 +226,19 @@ func (b *FixedSizeRingBuf) Write(p []byte) (n int, err error) {
 	}
 }
 
+// When there are no space to write, advance some space for write
+func (b *FixedSizeRingBuf) WriteAndMaybeOverwriteOldestData(p []byte) (n int, err error) {
+	writeCapacity := b.N - b.Readable
+	if len(p) > writeCapacity {
+		b.Advance(len(p) - writeCapacity)
+	}
+	startPos := 0
+	if len(p) > b.N {
+		startPos = len(p) - b.N
+	}
+	return b.Write(p[startPos:])
```

Comment:
It makes more sense to tell the client that all of len(p) was "written", as you remarked in the test code. So how about concluding with:
```
n, err = b.Write(p[startPos:])
if err != nil {
   return n, err
}
return len(p), nil
```